controller/volume_rebuilding_controller.go

```diff
@@ -192,12 +279,134 @@ func (vbc *VolumeRebuildingController) reconcile(volName string) (err error) {
 	}()
 
 	rebuildingAttachmentTicketID := longhorn.GetAttachmentTicketID(longhorn.AttacherTypeVolumeRebuildingController, volName)
+	if !vol.DeletionTimestamp.IsZero() {
+		vbc.logger.Infof("Volume %v is deleting, skip offline rebuilding", volName)
+		delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+		return nil
+	}
+	if vol.Status.Robustness == longhorn.VolumeRobustnessFaulted {
+		vbc.logger.Warnf("Volume %v is faulted, skip offline rebuilding", volName)
+		delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+		return nil
+	}
 
-	delete(va.Spec.AttachmentTickets, rebuildingAttachmentTicketID)
+	isOfflineRebuildEnabled, err := vbc.isVolumeOfflineRebuildEnabled(vol.Spec.OfflineRebuilding)
+	if err != nil {
+		return err
+	}
+	if isOfflineRebuildEnabled && !util.IsHigherPriorityVATicketExisting(va, longhorn.AttacherTypeVolumeRebuildingController) {
+		if vol.Status.State == longhorn.VolumeStateDetached {
+			va, err = vbc.syncLHVolumeAttachmentForOfflineRebuild(vol, va, rebuildingAttachmentTicketID)
+			if err != nil {
+				return err
+			}
+			return nil
+		}
+		if vol.Status.State == longhorn.VolumeStateAttaching {
+			return nil
+		}
+
+		isVolumeInRebuilding, err := vbc.isVolumeReplicasRebuilding(vol)
+		if err != nil {
+			return err
+		}
```

Comment:
Should we delete the volumeattachment ticket and return `nil` if the volume or engine is not found?

Reply:
>Should we delete the volumeattachment ticket and return nil if the volume or engine is not found?

Yes, if engine is `nil`, we could delete the va ticket as it occurred an unknown error.
Modified. Thanks.